// node-markov-generator/lib/textGenerator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fs = require("fs");
const tokenCollection_1 = require("./tokenCollection");
const sentenceHelper_1 = require("./sentenceHelper");
class TextGenerator {
    constructor(input) {
        this.wordsToStart = new tokenCollection_1.TokenCollection();
        this.wordsToFinish = new Set();
        this.wordStorage = new Map();
        let corpus;
        if (Array.isArray(input)) {
            corpus = input;
        }
        else {
            const corpusContents = fs.readFileSync(input).toString();
            corpus = corpusContents.split(/\r?\n/);
        }
        const minLastWordLength = 4;
        for (const line of corpus) {
            const sentences = sentenceHelper_1.SentenceHelper.splitIntoSentences(line.trim().toLowerCase()).map(s => s.trim())
                .filter(sentenceHelper_1.SentenceHelper.areAllCharsValid);
            for (const sentence of sentences) {
                const words = sentence.split(' ');
                if (!words.length) {
                    continue;
                }
                this.wordsToStart.add(words[0]);
                const lastWord = words[words.length - 1];
                if (lastWord.length >= minLastWordLength) {
                    this.wordsToFinish.add(lastWord);
                }
                for (let i = 0; i < words.length - 1; i++) {
                    const currentWord = words[i];
                    const nextWord = words[i + 1];
                    this.processWords(null, currentWord, nextWord);
                    if (i > 0) {
                        const previousWord = words[i - 1];
                        this.processWords(previousWord, currentWord, nextWord);
                    }
                }
            }
        }
    }
    processWords(prevWord, currentWord, nextWord) {
        const key = prevWord ? TextGenerator.getKeyForMultipleWords(prevWord, currentWord) : currentWord;
        if (this.wordStorage.has(key)) {
            this.wordStorage.get(key).add(nextWord);
        }
        else {
            this.wordStorage.set(key, new tokenCollection_1.TokenCollection([nextWord]));
        }
    }
    static getKeyForMultipleWords(...words) {
        return words.reduce((cur, next) => cur ? `${cur}|${next}` : next, '');
    }
    generate(options) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
        const minWordCount = (_b = (_a = options) === null || _a === void 0 ? void 0 : _a.minWordCount, (_b !== null && _b !== void 0 ? _b : 7));
        const maxWordCount = (_d = (_c = options) === null || _c === void 0 ? void 0 : _c.maxWordCount, (_d !== null && _d !== void 0 ? _d : 20));
        const contextUsageDegree = (_f = (_e = options) === null || _e === void 0 ? void 0 : _e.contextUsageDegree, (_f !== null && _f !== void 0 ? _f : 0.5));
        let retryCount = (_h = (_g = options) === null || _g === void 0 ? void 0 : _g.retryCount, (_h !== null && _h !== void 0 ? _h : 100));
        do {
            const wordToStart = (_l = (_k = (_j = options) === null || _j === void 0 ? void 0 : _j.wordToStart) === null || _k === void 0 ? void 0 : _k.toLowerCase(), (_l !== null && _l !== void 0 ? _l : this.wordsToStart.getRandom()));
            const result = this.generateInternal(wordToStart, minWordCount, maxWordCount, contextUsageDegree);
            retryCount--;
            if (result) {
                return result;
            }
        } while (retryCount);
        return null;
    }
    generateSentence(options) {
        var _a;
        const generated = this.generate(options);
        if ((_a = generated) === null || _a === void 0 ? void 0 : _a.length) {
            const [firstWord, ...rest] = generated;
            return [firstWord[0].toUpperCase() + firstWord.substr(1)].concat(rest).join(' ') + '.';
        }
        else {
            return null;
        }
    }
    generateInternal(wordToStart, minWordCount, maxWordCount, contextAwarenessDegree) {
        const resultWords = [wordToStart];
        let preLastGeneratedWord = null;
        let lastGeneratedWord = wordToStart;
        while (true) {
            const possibleNextWordAwareOfContext = preLastGeneratedWord ? this.wordStorage.get(TextGenerator.getKeyForMultipleWords(preLastGeneratedWord, lastGeneratedWord)) : null;
            const possibleSimpleNextWords = this.wordStorage.get(lastGeneratedWord);
            const possibleNextWords = possibleNextWordAwareOfContext && (Math.random() < contextAwarenessDegree) ? possibleNextWordAwareOfContext : possibleSimpleNextWords;
            if (!possibleNextWords) {
                if (resultWords.length >= minWordCount) {
                    return resultWords;
                }
                else {
                    return null;
                }
            }
            const nextWord = possibleNextWords.getRandom();
            if (nextWord) {
                resultWords.push(nextWord);
            }
            if (resultWords.length > minWordCount && this.wordsToFinish.has(nextWord)) {
                return resultWords;
            }
            if (resultWords.length === maxWordCount) {
                return null;
            }
            preLastGeneratedWord = lastGeneratedWord;
            lastGeneratedWord = nextWord;
        }
    }
}
exports.TextGenerator = TextGenerator;
